File: bindings/nodejs/src/path.cc
#include <napi.h>
#include <string>
#include <vector>
#include "../../../src/HBTrie/path.h"
#include "../../../src/HBTrie/identifier.h"
#include "../../../src/Buffer/buffer.h"

// Split string by delimiter
std::vector<std::string> SplitString(const std::string& str, char delimiter) {
  std::vector<std::string> tokens;
  size_t start = 0;
  size_t end = str.find(delimiter);

  while (end != std::string::npos) {
    tokens.push_back(str.substr(start, end - start));
    start = end + 1;
    end = str.find(delimiter, start);
  }
  tokens.push_back(str.substr(start));

  return tokens;
}

// Convert JavaScript key (string or array) to path_t*
path_t* PathFromJS(Napi::Env env, Napi::Value key, char delimiter) {
  path_t* path = path_create();
  if (!path) {
    Napi::Error::New(env, "Failed to create path").ThrowAsJavaScriptException();
    return nullptr;
  }

  if (key.IsString()) {
    std::string str = key.As<Napi::String>().Utf8Value();
    std::vector<std::string> parts = SplitString(str, delimiter);

    for (const auto& part : parts) {
      if (part.empty()) continue;  // Skip empty segments

      buffer_t* buf = buffer_create_from_pointer_copy(
        const_cast<uint8_t*>(reinterpret_cast<const uint8_t*>(part.c_str())),
        part.size()
      );
      if (!buf) {
        path_destroy(path);
        Napi::Error::New(env, "Failed to create buffer").ThrowAsJavaScriptException();
        return nullptr;
      }

      identifier_t* id = identifier_create(buf, 0);
      buffer_destroy(buf);
      if (!id) {
        path_destroy(path);
        Napi::Error::New(env, "Failed to create identifier").ThrowAsJavaScriptException();
        return nullptr;
      }

      path_append(path, id);
      identifier_destroy(id);
    }
  } else if (key.IsArray()) {
    Napi::Array arr = key.As<Napi::Array>();
    for (uint32_t i = 0; i < arr.Length(); i++) {
      Napi::Value part = arr.Get(i);
      if (!part.IsString()) {
        path_destroy(path);
        Napi::TypeError::New(env, "Path array elements must be strings").ThrowAsJavaScriptException();
        return nullptr;
      }

      std::string partStr = part.As<Napi::String>().Utf8Value();
      buffer_t* buf = buffer_create_from_pointer_copy(
        const_cast<uint8_t*>(reinterpret_cast<const uint8_t*>(partStr.c_str())),
        partStr.size()
      );
      if (!buf) {
        path_destroy(path);
        Napi::Error::New(env, "Failed to create buffer").ThrowAsJavaScriptException();
        return nullptr;
      }

      identifier_t* id = identifier_create(buf, 0);
      buffer_destroy(buf);
      if (!id) {
        path_destroy(path);
        Napi::Error::New(env, "Failed to create identifier").ThrowAsJavaScriptException();
        return nullptr;
      }

      path_append(path, id);
      identifier_destroy(id);
    }
  } else {
    path_destroy(path);
    Napi::TypeError::New(env, "Key must be string or array").ThrowAsJavaScriptException();
    return nullptr;
  }

  return path;
}

// Convert path_t* to JavaScript string
std::string PathToJS(path_t* path, char delimiter) {
  std::string result;

  for (size_t i = 0; i < path->identifiers.length; i++) {
    if (i > 0) {
      result += delimiter;
    }

    identifier_t* id = path->identifiers.data[i];

    // Extract identifier bytes to string
    for (size_t j = 0; j < id->chunks.length; j++) {
      chunk_t* chunk = id->chunks.data[j];
      const uint8_t* data = static_cast<const uint8_t*>(chunk_data_const(chunk));
      size_t size = chunk->data->size;

      // Check if all bytes are printable ASCII
      bool printable = true;
      for (size_t k = 0; k < size; k++) {
        if (!isprint(data[k]) && data[k] != '\t' && data[k] != '\n' && data[k] != '\r') {
          printable = false;
          break;
        }
      }

      if (printable) {
        result += std::string(reinterpret_cast<const char*>(data), size);
      } else {
        // Non-printable: represent as hex
        char hex[3];
        for (size_t k = 0; k < size; k++) {
          snprintf(hex, 3, "%02x", data[k]);
          result += hex;
        }
      }
    }
  }

  return result;
}

// Convert path_t* to JavaScript array
Napi::Array PathToArrayJS(Napi::Env env, path_t* path, char delimiter) {
  Napi::Array arr = Napi::Array::New(env);

  for (size_t i = 0; i < path->identifiers.length; i++) {
    identifier_t* id = path->identifiers.data[i];

    // Extract identifier as string
    std::string part;
    for (size_t j = 0; j < id->chunks.length; j++) {
      chunk_t* chunk = id->chunks.data[j];
      const uint8_t* data = static_cast<const uint8_t*>(chunk_data_const(chunk));
      size_t size = chunk->data->size;

      // Check if all bytes are printable ASCII
      bool printable = true;
      for (size_t k = 0; k < size; k++) {
        if (!isprint(data[k]) && data[k] != '\t' && data[k] != '\n' && data[k] != '\r') {
          printable = false;
          break;
        }
      }

      if (printable) {
        part += std::string(reinterpret_cast<const char*>(data), size);
      } else {
        // Non-printable: represent as hex
        char hex[3];
        for (size_t k = 0; k < size; k++) {
          snprintf(hex, 3, "%02x", data[k]);
          part += hex;
        }
      }
    }

    arr.Set(i, Napi::String::New(env, part));
  }

  return arr;
}

File: bindings/dart/lib/src/path.dart
// lib/src/path.dart
import 'dart:ffi';
import 'dart:convert';

import 'package:ffi/ffi.dart';

import 'native/types.dart';
import 'native/wavedb_bindings.dart';
import 'exceptions.dart';

/// Convert between Dart keys and native path_t
class PathConverter {
  /// Convert Dart key to native path_t
  ///
  /// Key can be:
  /// - String: Split by delimiter
  /// - List<String>: Direct path components
  static Pointer<path_t> toNative(dynamic key, String delimiter) {
    final path = WaveDBNative.pathCreate();

    try {
      List<String> parts;

      if (key is String) {
        // Split string by delimiter
        parts = key.split(delimiter).where((s) => s.isNotEmpty).toList();
      } else if (key is List) {
        // Already a list of path components
        parts = key.map((e) => e.toString()).toList();
      } else {
        throw ArgumentError('Key must be String or List<String>');
      }

      if (parts.isEmpty) {
        throw ArgumentError('Key cannot be empty');
      }

      // Create identifier for each path component and append
      for (final part in parts) {
        final id = _createIdentifierFromString(part);
        try {
          final rc = WaveDBNative.pathAppend(path, id);
          if (rc != 0) {
            throw WaveDBException.invalidPath(part, 'Failed to append to path');
          }
        } finally {
          // path_append increments refcount, does NOT take ownership
          // Must always destroy our reference after append
          WaveDBNative.identifierDestroy(id);
        }
      }

      return path;
    } catch (e) {
      // Clean up on error
      WaveDBNative.pathDestroy(path);
      rethrow;
    }
  }

  /// Create identifier from a string
  static Pointer<identifier_t> _createIdentifierFromString(String s) {
    // Use UTF-8 encoding, not UTF-16 code units
    final bytes = utf8.encode(s);
    final ptr = calloc<Uint8>(bytes.length);

    try {
      for (var i = 0; i < bytes.length; i++) {
        ptr[i] = bytes[i];
      }
      return WaveDBNative.identifierCreate(ptr, bytes.length);
    } finally {
      // identifierCreate copies the data, so we can free the buffer
      calloc.free(ptr);
    }
  }

  /// Convert native path_t to Dart key
  ///
  /// Returns String (joined by delimiter) or List<String> (if asArray=true)
  static dynamic fromNative(Pointer<path_t> path, String delimiter, {bool asArray = false}) {
    // TODO: Implement proper path reconstruction when FFI struct access is available
    // This requires reading the vec_t of identifiers from the path_t struct
    final parts = <String>[];

    if (asArray) {
      return parts;
    } else {
      return parts.join(delimiter);
    }
  }

  /// Destroy a native path
  static void destroy(Pointer<path_t> path) {
    WaveDBNative.pathDestroy(path);
  }
}
